main/src/com/google/refine/clustering/binning/FingerprintKeyer.java

```diff
@@ -69,222 +96,32 @@ public String key(String s, Object... o) {
         return b.toString();
     }
 
+    protected String normalize(String s) {
+        s = stripDiacritics(s);
+        s = stripNonDiacritics(s);
+        return s;
+    }
+
+    @Deprecated
     protected String asciify(String s) {
-        char[] c = s.toCharArray();
-        StringBuffer b = new StringBuffer();
-        for (char element : c) {
-            b.append(translate(element));
-        }
-        return b.toString();
+        return normalize(s);
     }
-    
-    /**
-     * Translate the given unicode char in the closest ASCII representation
-     * NOTE: this function deals only with latin-1 supplement and latin-1 extended code charts
-     */
-    private char translate(char c) {
```

Comment:
Random thought: to make sure we are not missing any of these cases by expressing the conversion with your regular expression, would it make sense to just transform this deleted code into a unit test which would check that all the behaviour is preserved for all the cases listed here?

Reply:
Good idea. I did that and updated the normalization with some substitutions for visually similar characters which are not normally considered the same. I also switched to the more powerful compatibility decomposition (NFKD) from the canonical decomposition (NFD) which is a bigger hammer, but this is intended to be a relatively blunt instrument. When we expose normalization as a function, we can allow more fine grained control.

The current code should handle all characters in the old table the same, except for O with stroke which is now "oe" instead of "o", and provide equivalent semantics for all characters and ligatures which were not included in the old table.

We need to decide (eventually) whether we want to continue the practice of aggressively folding Latin characters to their nearest ASCII equivalent even if the Java normalization code doesn't. The Latin Extended B and Latin Extended C characters (and probably others) would require us to do our own mappings. Latin Basic, Latin-1 Supplement, Latin Extended A and Latin Extended Additional are fully covered.